apps/api/src/deployment/services/top-up-custodial-deployments/top-up-custodial-deployments.service.ts

```diff
@@ -0,0 +1,136 @@
+  ) {}
+
+  async topUpDeployments() {
+    const wallets = [
+      { wallet: this.uaktMasterWalletService, client: this.uaktMasterSigningClientService },
+      { wallet: this.usdtMasterWalletService, client: this.usdtMasterSigningClientService }
+    ];
+
+    const topUpAllCustodialDeployments = wallets.map(async ({ wallet, client }) => {
+      const address = await wallet.getFirstAddress();
+      await this.allowanceHttpService.paginateDeploymentGrants({ grantee: address, limit: this.CONCURRENCY }, async grants => {
+        await Promise.all(
+          grants.map(async grant => {
+            await this.execWithErrorHandler(grant, () => this.topUpForGrant(grant, client));
+          })
+        );
+      });
+    });
+    await Promise.all(topUpAllCustodialDeployments);
+  }
+
+  private async topUpForGrant(grant: DeploymentAllowance, client: MasterSigningClientService) {
+    const owner = grant.granter;
+
+    const balances = await this.collectWalletBalances(grant);
+    this.logger.debug({ event: "BALANCES_COLLECTED", granter: owner, grantee: grant.grantee, balances });
+
+    const drainingDeployments = await this.retrieveDrainingDeployments(owner, balances.denom);
+    let { deploymentLimit, feesLimit, balance } = balances;
+
+    for (const deployment of drainingDeployments) {
+      const topUpAmount = await this.calculateTopUpAmount(deployment);
+      if (!this.canTopUp(topUpAmount, { deploymentLimit, feesLimit, balance })) {
+        this.logger.debug({ event: "INSUFFICIENT_BALANCE", granter: owner, grantee: grant.grantee, balances: { deploymentLimit, feesLimit, balance } });
+        break;
+      }
+      deploymentLimit -= topUpAmount;
+      feesLimit -= this.MIN_FEES_AVAILABLE;
+      balance -= topUpAmount + this.MIN_FEES_AVAILABLE;
+
+      await this.topUpDeployment(topUpAmount, deployment, client);
+    }
+  }
+
+  private async collectWalletBalances(grant: DeploymentAllowance): Promise<Balances> {
+    const denom = grant.authorization.spend_limit.denom;
+    const deploymentLimit = parseFloat(grant.authorization.spend_limit.amount);
+
+    const feesLimit = await this.retrieveFeesLimit(grant.granter, grant.grantee, denom);
+    const { amount } = await this.balanceHttpService.getBalance(grant.granter, denom);
+    const balance = parseFloat(amount);
+
+    return {
+      denom,
+      feesLimit,
+      deploymentLimit,
+      balance
+    };
+  }
+
+  private async retrieveFeesLimit(granter: string, grantee: string, denom: string) {
+    const feesAllowance = await this.allowanceHttpService.getFeeAllowanceForGranterAndGrantee(granter, grantee);
+    const feesSpendLimit = feesAllowance.allowance.spend_limit.find(limit => limit.denom === denom);
+
+    return feesSpendLimit ? parseFloat(feesSpendLimit.amount) : 0;
+  }
+
+  private async retrieveDrainingDeployments(owner: string, denom: string): Promise<DrainingDeploymentOutput[]> {
+    const currentHeight = await this.blockHttpService.getCurrentHeight();
```

Comment:
This does an axios call to a node and it is being called concurrently for multiple deployments. Should we add caching since it will most likely return de same value for every calls?

Reply:
yea, i initially added this for simplicity but forgot to change

Reply:
https://github.com/akash-network/console/pull/439/files#diff-4f235eb0e292279ed8db718ddd636e379b7bf3aa6c8858de1a9234099fdb89abR116